plutus-emulator/src/Wallet/Emulator/NodeClient.hs

```diff
@@ -0,0 +1,150 @@
+    _index            :: Index.UtxoIndex -- ^ The UTxO index, used for validation.
+} deriving (Show)
+
+makeLenses ''NodeClientState
+
+-- | Validate a block given the current slot and UTxO index, returning the valid
+--   transactions, success/failure events, remaining transactions and the
+--   updated UTxO set.
+validateBlock :: Slot -> Index.UtxoIndex -> [Tx] -> (ValidatedBlock, Index.UtxoIndex)
+validateBlock slot@(Slot s) idx txns =
+    let
+        -- Select those transactions that can be validated in the
+        -- current slot
+        (eligibleTxns, rest) = partition (canValidateNow slot) txns
+
+        -- Validate eligible transactions, updating the UTXO index each time
+        (processed, idx') =
+            flip runState idx $ for eligibleTxns $ \t -> do
+                r <- validateEm slot t
+                pure (t, r)
+
+        -- The new block contains all transaction that were validated
+        -- successfully
+        block = fst <$> filter (isNothing . snd) processed
+
+        -- Also return an `EmulatorEvent` for each transaction that was
+        -- processed
+        nextSlot = Slot (s + 1)
+        events   = SlotAdd nextSlot : (uncurry mkEvent <$> processed)
+
+    in (ValidatedBlock block events rest, idx')
+
+-- | Check whether the given transaction can be validated in the given slot.
+canValidateNow :: Slot -> Tx -> Bool
+canValidateNow slot tx = Interval.member slot (txValidRange tx)
+
+mkEvent :: Tx -> Maybe Index.ValidationError -> NodeClientEvent
+mkEvent t result =
+    case result of
+        Nothing  -> TxnValidate (hashTx t)
+        Just err -> TxnValidationFail (hashTx t) err
+
+-- | Validate a transaction in the current emulator state.
+validateEm :: MonadState Index.UtxoIndex m => Slot -> Tx -> m (Maybe Index.ValidationError)
+validateEm h txn = do
+    idx <- get
+    let result = Index.runValidation (Index.validateTransaction h txn) idx
+    case result of
+        Left e -> pure (Just e)
+        Right idx' -> do
+            _ <- put idx'
+            pure Nothing
+
+class MonadState NodeClientState m => NodeClientAPI m where
+    publishTx :: Tx -> m ()
+    currentSlot :: m Slot
+    eventStream :: Slot -> ConduitT () Block m ()
+    processBlock :: m ValidatedBlock
+            
+-- | Pure implementation of the NodeClient API
+
+instance NodeClientAPI (State NodeClientState) where 
+    publishTx tx = state $ ((), ) . over txPool (tx :)
+
+    currentSlot = 
+        (Slot . fromIntegral . length . _chainNewestFirst) <$> get
+
+    eventStream slot = loop slot
+        where
+            loop :: MonadState NodeClientState m
+                 => Slot 
+                 -> ConduitT () Block m ()
+            loop s = do
+                blocks <-
+                    drop (fromIntegral s) . _chainNewestFirst <$> get
+                void $ yieldMany blocks
+                loop $ s + fromIntegral (length blocks)
+
+    processBlock = do
```

Comment:
Could this just be a function with a `MonadState NodeState m` constraint?

Reply:
i.e. we're not really expecting that different class instances are going to have different implementations of this function, are we?

Reply:
Initially, everything was implemented using functions, but we decided to go with a monad to make clear what the API is. Since this method is part of the negotiated APIs, I think we should keep it.

Reply:
I'm not sure about that. *Derived* operations should be fine to put in methods. Generally I would say that if you aren't going to have different implementations for different class instances, then it shouldn't be a class method!